src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp

```diff
@@ -79,60 +93,78 @@ void WriteBufferFromAzureBlobStorage::finalizeImpl()
 {
     execWithRetry([this](){ next(); }, max_unexpected_write_error_retries);
 
-    if (tmp_buffer_write_offset > 0)
-        uploadBlock(tmp_buffer->data(), tmp_buffer_write_offset);
+    task_tracker->waitAll();
 
     auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
     execWithRetry([&](){ block_blob_client.CommitBlockList(block_ids); }, max_unexpected_write_error_retries);
 
     LOG_TRACE(log, "Committed {} blocks for blob `{}`", block_ids.size(), blob_path);
 }
 
-void WriteBufferFromAzureBlobStorage::uploadBlock(const char * data, size_t size)
+void WriteBufferFromAzureBlobStorage::nextImpl()
 {
-    auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
-    const std::string & block_id = block_ids.emplace_back(getRandomASCIIString(64));
+    task_tracker->waitIfAny();
 
-    Azure::Core::IO::MemoryBodyStream memory_stream(reinterpret_cast<const uint8_t *>(data), size);
-    execWithRetry([&](){ block_blob_client.StageBlock(block_id, memory_stream); }, max_unexpected_write_error_retries, size);
-    tmp_buffer_write_offset = 0;
+    reallocateBuffer();
```

Comment:
The call of `reallocateBuffer()` looks wrong here. `nextImpl()` is called when the buffer is already filled with data, why do you need to reallocate the buffer at that point?

Reply:
Please remove the call `reallocateBuffer()` here, it seems to be redundant.

Reply:
addressed it